feedback.go

```diff
@@ -0,0 +1,116 @@
+package apns
+
+import (
+	"crypto/tls"
+	"encoding/binary"
+	"encoding/hex"
+	"io"
+	"strings"
+	"time"
+
+	"code.google.com/p/go.net/context"
+)
+
+const (
+	FeedbackGateway        string = "feedback.push.apple.com:2196"
+	FeedbackSandboxGateway string = "feedback.sandbox.push.apple.com:2196"
+)
+
+const (
+	feedbackCheckPeriod time.Duration = 5 * time.Second
+)
+
+type FeedbackMessage struct {
+	Unsubscribe time.Time
+	DeviceToken string
+}
+
+// Feedback only read binary apple socket
+type Feedback struct {
+	addr     string
+	cert     *tls.Certificate
+	messages chan *FeedbackMessage
+}
+
+func NewFeedback(ctx context.Context, addr string, cert *tls.Certificate) (f *Feedback) {
+	f = &Feedback{
+		addr:     addr,
+		cert:     cert,
+		messages: make(chan *FeedbackMessage),
+	}
+	go f.reader(ctx)
+	return
+}
+
+func (f *Feedback) Messages() <-chan *FeedbackMessage {
+	return f.messages
+}
+
+func (f *Feedback) reader(ctx context.Context) {
+	var stop bool
+	for {
+		if stop {
+			break
+		}
+		result, err := f.receive()
+		if err != nil && err != io.EOF {
+			info("Feedback receive err: %v", err)
+		} else {
+			for _, msg := range result {
+				info("Feedback receive msg: %v", msg)
+				f.messages <- msg
+			}
+		}
+		time.Sleep(feedbackCheckPeriod)
+	}
+
+	go func() {
```

Comment:
It seems only the goroutine on line 67 can set `stop` to true, however the goroutine is never going to be executed ;)